### PR TITLE
[BEAM-8337] Hard-code Flink versions.

### PR DESCRIPTION
**Pull request by ibzib**
Also publish the RC images instead of rebuilding.

There were problems with the bash code for munging Flink versions, so I got rid of it and replaced with a hard-coded list. The release manager will have to manually validate that they're correct, but I think that's a good idea anyway.

@Hannah-Jiang @udim @mxm 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
> What were the problems? I'd prefer not hardcoding the versions. If there are problems with this, I'd reckon they are just a symptom of other non-deterministic patterns in the release scripts.

The issues were with creating FLINK_VER as a bash array from subshell output. I recommend using strings instead. I don't mind hard-coded versions for now.

File: release/src/main/scripts/publish_docker_images.sh [hunk 1]
```diff
@@ -97,18 +97,26 @@ if [[ $confirmation = "y" ]]; then
...
+  read
   for ver in "${FLINK_VER[@]}"; do
-     ./gradlew ":runners:flink:${ver}:job-server-container:docker" -Pdocker-tag="${RELEASE}"
      FLINK_IMAGE_NAME=apachebeam/flink${ver}_job_server
+
+     # Pull verified RC from dockerhub.
+     docker pull "${FLINK_IMAGE_NAME}:${RELEASE}_${RC_VERSION}"
+
+     # Tag with ${RELEASE} and push to dockerhub.
+     docker tag "${FLINK_IMAGE_NAME}:${RELEASE}_${RC_VERSION}" "${FLINK_IMAGE_NAME}:${RELEASE}"
```
Comment (mxm) [inline release/src/main/scripts/publish_docker_images.sh]: Looks like the indention is off here (3 instead of 2 spaces).

**Comment by ibzib** [comment]
We could've fixed the bash instead. I prefer this way because the publish
script no longer has to either make a clone or be run from within an up to
date clone, instead it can be run from anywhere.

**Comment by Hannah-Jiang** [comment]
Looks good to me.
One comment is, do we need to update [release guide](https://github.com/apache/beam/blob/master/website/src/contribute/release-guide.md) as well?

**Comment by ibzib** [comment]
> One comment is, do we need to update release guide as well?

I see that the scripts are also written in markdown within the release guide ["run all steps manually"](https://github.com/apache/beam/blob/master/website/src/contribute/release-guide.md#alternative-run-all-steps-manually-1) section. I'm not sure why that section's there, however. Is there a difference between that and the contents of the scripts?

**Comment by Hannah-Jiang** [comment]
> > One comment is, do we need to update release guide as well?
> 
> I see that the scripts are also written in markdown within the release guide ["run all steps manually"](https://github.com/apache/beam/blob/master/website/src/contribute/release-guide.md#alternative-run-all-steps-manually-1) section. I'm not sure why that section's there, however. Is there a difference between that and the contents of the scripts?

This script is used when release it manually, for example, the script failed during running and release manager release left part manually. So we need to keep it consistent. However, I think this part can be improved.

**Comment by ibzib** [comment]
> So we need to keep it consistent

I'd prefer to keep only the full scripts as the single source-of-truth for what needs to be run. But that's out of scope for this PR.

**Comment by ibzib** [comment]
For now, I've updated the release guide. For future discussion/action on simplifying the release guide, I filed https://issues.apache.org/jira/browse/BEAM-9087.

**Pull request merged**